### ISSUE TITLE
slow inference time when use aot to cogvideo5b

### ISSUE THREAD

**Issue by DamonsJ** [sayakpaul/diffusers-torchao#41]
this is my code:
```
import numpy as np
import torch
from diffusers import (AutoencoderKL, CogVideoXDDIMScheduler, DDIMScheduler,
                       DPMSolverMultistepScheduler,
                       EulerAncestralDiscreteScheduler, EulerDiscreteScheduler,
                       PNDMScheduler)
from transformers import T5EncoderModel, T5Tokenizer
from omegaconf import OmegaConf
from PIL import Image

from cogvideox.models.transformer3d import CogVideoXTransformer3DModel
from cogvideox.models.autoencoder_magvit import AutoencoderKLCogVideoX
from cogvideox.pipeline.pipeline_cogvideox import CogVideoX_Fun_Pipeline
from cogvideox.pipeline.pipeline_cogvideox_inpaint import CogVideoX_Fun_Pipeline_Inpaint
from cogvideox.utils.lora_utils import merge_lora, unmerge_lora
from cogvideox.utils.utils import get_image_to_video_latent, save_videos_grid

import torch.utils.benchmark as benchmark
from functools import partial

def get_example_inputs():
    example_inputs = {
        "hidden_states": torch.randn(2, 13, 16,128, 128, dtype=torch.bfloat16, device="cuda"),
        "encoder_hidden_states": torch.randn(2, 226, 4096, dtype=torch.bfloat16, device="cuda"),
        "timestep":  torch.randint(0,4096,(2,), device="cuda"),
        "inpaint_latents":torch.randn(2, 13, 17,128, 128, dtype=torch.bfloat16, device="cuda"),
        "image_rotary_emb": ((torch.randn(64, dtype=torch.bfloat16, device="cuda"),torch.randn(64, dtype=torch.bfloat16, device="cuda"))),
        "timestep_cond":None,
        "control_latents":None,
        "return_dict":False,
    }
    return example_inputs

def benchmark_fn(f, *args, **kwargs):
    t0 = benchmark.Timer(
        stmt="f(*args, **kwargs)",
        globals={"args": args, "kwargs": kwargs, "f": f},
        num_threads=torch.get_num_threads(),
    )
    return f"{(t0.blocked_autorange().mean):.3f}"

def load_model():
    weight_dtype = torch.bfloat16
    transformer = CogVideoXTransformer3DModel.from_pretrained_2d(
        "/yourpath", 
        subfolder="transformer",
    ).to(weight_dtype)
    transformer.to("cuda")
    return transformer

def aot_compile(name, fn, **sample_kwargs):
    path = f"{name}.so"
    print(f"{path=}")
    options = {
        "aot_inductor.output_path": path,
        "max_autotune": True,
        "triton.cudagraphs": True,
    }

    torch._export.aot_compile(
        fn,
        (),
        sample_kwargs,
        options=options,
        disable_constraint_solver=True,
    )
    return path 

def aot_load(path):
    return torch._export.aot_load(path, "cuda")

@torch.no_grad()
def f(model, **kwargs):
    return model(**kwargs)

model = load_model()

inputs = get_example_inputs()
path = aot_compile("bs_1_1024", partial(f, model=model), **inputs)
compiled_func_1 = aot_load(path)
print(f"{compiled_func_1(**inputs)[0].shape=}")

for _ in range(5):
    _ = compiled_func_1(**inputs)[0]

time = benchmark_fn(f, compiled_func_1, **inputs)
print("aot time is : ", time) 

transformer_model = load_model()
torch._inductor.config.conv_1x1_as_mm = True
torch._inductor.config.coordinate_descent_tuning = True
torch._inductor.config.epilogue_fusion = False
torch._inductor.config.coordinate_descent_check_all_directions = True
transformer_model.to(memory_format=torch.channels_last)
transformer_model = torch.compile(transformer_model, mode="max-autotune", fullgraph=True)
time = benchmark_fn(f, transformer_model, **inputs)
time = benchmark_fn(f, transformer_model, **inputs)
print("jit compile time is : ", time) 

```

this is the info printed:

```
compiled_func_1(**inputs)[0].shape=torch.Size([2, 13, 16, 128, 128])
aot time is :  11.705
### missing keys: 0; 
### unexpected keys: 0;
[]
### Mamba Parameters: 0.0 M
### attn1 Parameters: 1585.97376 M
jit compile time is :  11.646
```
inference time is almost equal

so, does aot compile help to decrease inference time? not for all model ?

**Comment by sayakpaul** [comment]
Cc: @cpuhrsch

**Comment by cpuhrsch** [comment]
@DamonsJ which GPU are you running this on?

**Comment by DamonsJ** [comment]
Wed Dec  4 03:33:09 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:CB:00.0 Off |                   On |
| N/A   34C    P0             74W /  700W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| MIG devices:                                                                            |
+------------------+----------------------------------+-----------+-----------------------+
| GPU  GI  CI  MIG |                     Memory-Usage |        Vol|      Shared           |
|      ID  ID  Dev |                       BAR1-Usage | SM     Unc| CE ENC DEC OFA JPG    |
|                  |                                  |        ECC|                       |
|==================+==================================+===========+=======================|
|  0    0   0   0  |              97MiB / 80995MiB    |132      0 |  8   0    7    1    7 |
|                  |                 1MiB / 131072MiB |           |                       |
+------------------+----------------------------------+-----------+-----------------------+

**Comment by DamonsJ** [comment]
@cpuhrsch 

![image](https://github.com/user-attachments/assets/cfd0517a-7eed-464b-904c-1ce95156c862)

**Comment by cpuhrsch** [comment]
Hm, interesting. Can you try generating a GPU trace as in https://github.com/pytorch-labs/segment-anything-fast/tree/main/experiments#kernel-traces ? Just make sure the trace ends in `.json.gz`.

**Comment by DamonsJ** [comment]
@cpuhrsch  hi, this is the files generated :
[cogvideo-transformer-jit.json.gz](https://github.com/user-attachments/files/18005778/cogvideo-transformer-jit.json.gz)
[cogvideo-transformer-aot.json.gz](https://github.com/user-attachments/files/18005779/cogvideo-transformer-aot.json.gz)

and the code I used is :

```
import numpy as np
import torch
from diffusers import (AutoencoderKL, CogVideoXDDIMScheduler, DDIMScheduler,
                       DPMSolverMultistepScheduler,
                       EulerAncestralDiscreteScheduler, EulerDiscreteScheduler,
                       PNDMScheduler)
from transformers import T5EncoderModel, T5Tokenizer
from omegaconf import OmegaConf
from PIL import Image

from cogvideox.models.transformer3d import CogVideoXTransformer3DModel
from cogvideox.models.autoencoder_magvit import AutoencoderKLCogVideoX
from cogvideox.pipeline.pipeline_cogvideox import CogVideoX_Fun_Pipeline
from cogvideox.pipeline.pipeline_cogvideox_inpaint import CogVideoX_Fun_Pipeline_Inpaint
from cogvideox.utils.lora_utils import merge_lora, unmerge_lora
from cogvideox.utils.utils import get_image_to_video_latent, save_videos_grid

import torch.utils.benchmark as benchmark
from functools import partial

def get_example_inputs():
    example_inputs = {
        "hidden_states": torch.randn(2, 13, 16,128, 128, dtype=torch.bfloat16, device="cuda"),
        "encoder_hidden_states": torch.randn(2, 226, 4096, dtype=torch.bfloat16, device="cuda"),
        "timestep":  torch.randint(0,4096,(2,), device="cuda"),
        "inpaint_latents":torch.randn(2, 13, 17,128, 128, dtype=torch.bfloat16, device="cuda"),
        "image_rotary_emb": ((torch.randn(64, dtype=torch.bfloat16, device="cuda"),torch.randn(64, dtype=torch.bfloat16, device="cuda"))),
        "timestep_cond":None,
        "control_latents":None,
        "return_dict":False,
    }
    return example_inputs

def benchmark_fn(f, *args, **kwargs):
    t0 = benchmark.Timer(
        stmt="f(*args, **kwargs)",
        globals={"args": args, "kwargs": kwargs, "f": f},
        num_threads=torch.get_num_threads(),
    )
    return f"{(t0.blocked_autorange().mean):.3f}"

def load_model():
    weight_dtype = torch.bfloat16
    transformer = CogVideoXTransformer3DModel.from_pretrained_2d(
        "/workspace/cogvideox-fun/models/CogVideoX-Fun-V1.1-5b-InP",
        subfolder="transformer",
    ).to(weight_dtype)
    transformer.to("cuda")
    return transformer

def aot_compile(name, fn, **sample_kwargs):
    path = f"{name}.so"
    print(f"{path=}")
    options = {
        "aot_inductor.output_path": path,
        "max_autotune": True,
        "triton.cudagraphs": True,
    }

    torch._export.aot_compile(
        fn,
        (),
        sample_kwargs,
        options=options,
        disable_constraint_solver=True,
    )
    return path 

def aot_load(path):
    return torch._export.aot_load(path, "cuda")

@torch.no_grad()
def f(model, **kwargs):
    return model(**kwargs)

model = load_model()
inputs = get_example_inputs()
path = aot_compile("bs_1_1024", partial(f, model=model), **inputs)
path = "/workspace/cogvideox-fun/bs_1_1024.so"
compiled_func_1 = aot_load(path)
print(f"{compiled_func_1(**inputs)[0].shape=}")

for _ in range(5):
    _ = compiled_func_1(**inputs)[0]
with torch.profiler.profile(
            activities=[torch.profiler.ProfilerActivity.CPU,
                        torch.profiler.ProfilerActivity.CUDA],
            record_shapes=True) as prof:
    time = benchmark_fn(f, compiled_func_1, **inputs)
prof.export_chrome_trace("cogvideo-transformer-aot.json.gz")
print("aot time is : ", time) 

transformer_model = load_model()
torch._inductor.config.conv_1x1_as_mm = True
torch._inductor.config.coordinate_descent_tuning = True
torch._inductor.config.epilogue_fusion = False
torch._inductor.config.coordinate_descent_check_all_directions = True
transformer_model.to(memory_format=torch.channels_last)
transformer_model = torch.compile(transformer_model, mode="max-autotune", fullgraph=True)
time = benchmark_fn(f, transformer_model, **inputs)
with torch.profiler.profile(
            activities=[torch.profiler.ProfilerActivity.CPU,
                        torch.profiler.ProfilerActivity.CUDA],
            record_shapes=True) as prof:
    time = benchmark_fn(f, transformer_model, **inputs)
prof.export_chrome_trace("cogvideo-transformer-jit.json.gz")
print("jit compile time is : ", time) 


```

**Comment by cpuhrsch** [comment]
Hm, interesting. Looking at the trace it is very dominated by scaled dot product attention (flash attention). Maybe it's just in the nature of this model. If one operation entirely dominates the overall runtime, then speeding up the other operations won't help much. Note that we've released a CUDNN backends for SDPA that (without looking at the specifics of the model) maybe help here: https://github.com/pytorch/pytorch/releases/tag/v2.5.0

**Comment by DamonsJ** [comment]
@cpuhrsch  Thanks very much !
The above code already run on pytorch 2.5.1
It seems that the new SDPA CUDNN backends is default.
I don't know how to check

**Comment by cpuhrsch** [comment]
@DamonsJ - Ah, you can see if it's possible to run by using the context manager. See https://pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html#torch.nn.attention.sdpa_kernel and https://pytorch.org/docs/stable/generated/torch.nn.attention.SDPBackend.html#torch.nn.attention.SDPBackend for the choices. Furthermore we also have that

"""
Each of the fused kernels has specific input limitations. If the user requires the use of a specific fused implementation, disable the PyTorch C++ implementation using [torch.nn.attention.sdpa_kernel()](https://pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html#torch.nn.attention.sdpa_kernel). In the event that a fused implementation is not available, a warning will be raised with the reasons why the fused implementation cannot run.
"""

So if you do

```
with sdpa_kernel(backends=[SDPBackend.CUDNN_ATTENTION]):
    model(...)
```

It'll constrain the whole model to only use CUDNN and also tell you why it can't use it (e.g. head dim constraints etc.)

Having said all of this, if you're using the most recent PyTorch then we should already pick the best backend for you automatically. Upcoming versions will continue to improve the performance of SDPA. Depending on the specifics of the attention used in cog you might also want to give https://pytorch.org/blog/flexattention a try.

**Comment by DamonsJ** [comment]
@cpuhrsch  Thanks very much!
I will try!

**Comment by sayakpaul** [comment]
Closing due to inactivity.
